src/main/java/org/reaktivity/nukleus/tcp/internal/writer/stream/Slab.java

```diff
@@ -0,0 +1,170 @@
+
+    Slab(int maximumPartiallyWrittenStreams, int maximumWindowSize)
+    {
+        validateGT0(maximumPartiallyWrittenStreams, "maximumPartiallyWrittenStreams");
+        validateGT0(maximumWindowSize, "maximumWindowSize");
+        int slots = findNextPositivePowerOfTwo(maximumPartiallyWrittenStreams);
+        slotSize = findNextPositivePowerOfTwo(maximumWindowSize);
+        slotSizeBits = Integer.numberOfTrailingZeros(slotSize);
+        int capacityInBytes = slots << slotSizeBits;
+        buffer = ByteBuffer.allocateDirect(capacityInBytes);
+        writeBuffer = ByteBuffer.allocateDirect(slotSize);
+        offsets = new int[slots];
+        Arrays.fill(offsets, UNUSED);
+        remaining = new int[slots];
+    }
+
+    public ByteBuffer get(int slot, DirectBuffer data, int offset, int length)
+    {
+        ByteBuffer result;
+        if (slot == NO_SLOT)
+        {
+            writeBuffer.clear();
+            data.getBytes(offset, writeBuffer, length);
+            writeBuffer.flip();
+            result = writeBuffer;
+        }
+        else
+        {
+            // Append the data to the previous remaining data
+            buffer.position(offsets[slot] + remaining[slot]);
+            buffer.limit((slot << slotSizeBits) + slotSize);
+            data.getBytes(offset, buffer, length);
+            remaining[slot] += length;
+            buffer.position(offsets[slot]);
+            buffer.limit(buffer.position() + remaining[slot]);
+            result = buffer;
+        }
+        return result;
+    }
+
+    public ByteBuffer get(int slot)
+    {
+        ByteBuffer result = null;
+        if (slot != NO_SLOT)
+        {
+            buffer.position(offsets[slot]);
+            buffer.limit(offsets[slot] + remaining[slot]);
+            result = buffer;
+        }
+        return result;
+    }
+
+    public int written(long streamId, int slot, ByteBuffer written)
+    {
+        int nextSlot = slot;
+        if (slot == NO_SLOT)
+        {
+            if (written.hasRemaining())
+            {
+                // store the remaining data into a new slot
+                nextSlot = acquire(streamId);
+                if (nextSlot != OUT_OF_MEMORY)
+                {
+                    remaining[nextSlot] = written.remaining();
+                    buffer.limit(offsets[nextSlot] + slotSize);
+                    buffer.position(offsets[nextSlot]);
+                    buffer.put(written);
+                    buffer.limit(buffer.position());
+                    buffer.position(offsets[nextSlot]);
+                }
+            }
+            // else: nothing to do, continue to use the default writeBuffer
+        }
+        else
+        {
+            if (written.hasRemaining())
+            {
+                // Some data from the existing slot was written, adjust offset and remaining
+                offsets[slot] = written.position();
+                remaining[slot] = written.remaining();
```

Comment:
What happens in the following scenario? (assumes slab slots size 8192)
- BEGIN received, WINDOW [update=8192] sent
window size now 8192 (initial)
slab slot not used yet
- DATA [length=2048] received, write only 1024 bytes, so 1024 remaining on slot in slab
window size now 6144 = 8192 - 2048 (bytes received)
slab slot using offset 0 length 1024
- WINDOW [update=1024] sent for 1024 written bytes
window size now 7168 = 6144 + 1024 (bytes written)
- DATA [length=2048] received, added to slab slot
window size now 5020 = 7168 - 2048 (bytes received)
slab slot using offset 0 length 3072
- OP_WRITE triggered, 2048 bytes written, offset in Slab slot now 2048
slab slot using offset 2048 length 1024
- WINDOW [update=2048] sent for 2048 written bytes
window size now 7168 = 5020 + 2048 (bytes written)
- DATA [length=6144] received, added to slab slot at offset 3072
window size now 1024 = 7168 - 6144 (bytes received)
slab slot using offset 2048 length 7168

Wouldn't the last DATA frame cause the slab slot to overflow past 8192 due to 2048 + 7168 = 9216 > 8192 ?